File: owens_t/owens_t__.cpp
// mex instructions:
// mex owens_t__.cpp -I'C:\boost_1_87_0' -output owens_t__

#include "mexAdapter.hpp"
#include <stdexcept>
#include <boost/math/special_functions/owens_t.hpp>

class MexFunction : public matlab::mex::Function {
public:
    void operator()(matlab::mex::ArgumentList outputs, matlab::mex::ArgumentList inputs) {
        
        // Ensure 2 inputs
        if (inputs.size() != 2)
            throw std::invalid_argument( "Must pass 2 arguments" );

        // Get TypedArray<double> pointers to input arrays
        matlab::data::TypedArray<double> h = std::move(inputs[0]);
        matlab::data::TypedArray<double> a = std::move(inputs[1]);

        // Return if either input is empty
        if ( h.isEmpty() || a.isEmpty() )
            throw std::invalid_argument( "Empty array passed" );

        // Get array sizes
        size_t n = h.getNumberOfElements();
        size_t m = a.getNumberOfElements();
        
        // Ensure matching sizes of h and a or that at least one is scalar
        if (  !(n == m || n == 1 || m == 1)  )
            throw std::invalid_argument( "Input array size mismatch" );

        // Allocate memory for output
        matlab::data::ArrayFactory f;
        matlab::data::TypedArray<double> y = f.createArray<double>(std::vector<size_t>{std::max(n,m)});

        // Cast MATLAB TypedArrays to STL
        // TypedArray<double> -> unique_ptr<double[]>
        auto H = h.release();
        auto A = a.release();

        // Define data policy
        typedef boost::math::policies::policy<
            boost::math::policies::promote_double<false>,
            boost::math::policies::promote_float<false>,
            boost::math::policies::max_series_iterations<100>
            > my_policy;

        // Compute output
        if (n == m) {
            for(size_t i = 0; i < n; i++)
                y[i] = boost::math::owens_t(H[i], A[i], my_policy());
        }
        else if (m == 1) {
            for(size_t i = 0; i < n; i++)
                y[i] = boost::math::owens_t(H[i], A[0], my_policy());
        }
        else { // if (n == 1)
            for(size_t i = 0; i < m; i++)
                y[i] = boost::math::owens_t(H[0], A[i], my_policy());
        }    
          
        // Return output
        outputs[0] = y;
    }
};

File: owens_t/owens_t.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
% Owen's T function, T(h, a), gives the probability of the event 
%   (X > h and 0 < Y < aX)
% where X and Y are independent standard normal random variables.
%
% Defined in
% Owen, D B (1956). "Tables for computing bivariate normal probabilities".
%       Annals of Mathematical Statistics, 27, 1075–1090.
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
% USAGE
%   p = owens_t(h,a);
%
% INPUT
%   h - Array of any number of dimensions. Must be either scalar or contain
%       the same number of elements as 'a'.
%   a - Array of any number of dimensions. Must be either scalar or contain
%       the same number of elements as 'h'.
%
% OUTPUT
%   p - Array with the same size as 'h' or 'a'.
%
% EXCEPTIONS
%   (1) Fewer or more than 2 arguments were passed.
%           MESSAGE: "Must pass 2 arguments."
%
%   (2) 'h' or 'a' are not type double.
%           MESSAGE: "Can't convert the Array to this TypedArray"
%
%   (3) 'h' or 'a' are empty.
%           MESSAGE: "One or more empty argument(s) passed."
%
%   (4) Size mismatch between 'h' and 'a'. They must be the same size or at
%       least one must be scalar.
%           MESSAGE: "Input size mismatch. Arguments must be either the same size or at least one argument must be scalar."

% COMPILATION
%   This is a wrapper around the boost::math::owens_t() C++ function to
%   port an Owen's T function into MATLAB. The code is architecture
%   independent, so it should recompile on any system.
% 
%   Current version compiled under
%       MSVC    19.40.33813
%       boost   1.87.0
%
%   Compilation instructions:
%       (1) Boost libraries are available for download at https://www.boost.org/
%       (2) Ensure MEX is configured for C++ language compilation. To
%           verify, type the following in the Command Window:
%               >> mex -setup c++
%       (3) In MATLAB, set working directory to location of owens_t.cpp
%       (4) Type in the Command Window:
%               >> mex owens_t.cpp -I'<boost_parent_directory>' -output owens_t
%           where <boost_parent_directory> is the top directory of boost.
%           Example:
%               >> mex owens_t.cpp -I'C:\boost_1_87_0' -output owens_t
%
%
% AUTHOR:
%   Devin H. Kehoe
%   dhkehoe@gmail.com
%   April 24, 2025